aea/test_tools/docker_image.py

```diff
@@ -0,0 +1,204 @@
+import logging
+import os
+import re
+import shutil
+import subprocess  # nosec
+import tempfile
+import time
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional
+
+import docker
+import pytest
+from docker import DockerClient
+from docker.models.containers import Container
+
+from aea.exceptions import enforce
+from aea.helpers import http_requests as requests
+
+
+logger = logging.getLogger(__name__)
+
+
+class DockerImage(ABC):
+    """A class to wrap interatction with a Docker image."""
+
+    MINIMUM_DOCKER_VERSION = (19, 0, 0)
+
+    def __init__(self, client: docker.DockerClient):
+        """Initialize."""
+        self._client = client
+
+    def check_skip(self) -> None:
+        """
+        Check whether the test should be skipped.
+
+        By default, nothing happens.
+        """
+        self._check_docker_binary_available()
+
+    def _check_docker_binary_available(self) -> None:
+        """Check the 'Docker' CLI tool is in the OS PATH."""
+        result = shutil.which("docker")
+        if result is None:
+            pytest.skip("Docker not in the OS Path; skipping the test")
+
+        result = subprocess.run(  # nosec
+            ["docker", "--version"], stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        if result.returncode != 0:
+            pytest.skip(f"'docker --version' failed with exit code {result.returncode}")
+
+        match = re.search(
+            r"Docker version ([0-9]+)\.([0-9]+)\.([0-9]+)",
+            result.stdout.decode("utf-8"),
+        )
+        if match is None:
+            pytest.skip("cannot read version from the output of 'docker --version'")
+        version = (int(match.group(1)), int(match.group(2)), int(match.group(3)))
+        if version < self.MINIMUM_DOCKER_VERSION:
+            pytest.skip(
+                f"expected Docker version to be at least {'.'.join(self.MINIMUM_DOCKER_VERSION)}, found {'.'.join(version)}"
+            )
+
+    @property
+    @abstractmethod
+    def tag(self) -> str:
+        """Return the tag of the image."""
+
+    def stop_if_already_running(self) -> None:
+        """Stop the running images with the same tag, if any."""
+        client = docker.from_env()
+        for container in client.containers.list():
+            if self.tag in container.image.tags:
+                logger.info(f"Stopping image {self.tag}...")
+                container.stop()
+
+    @abstractmethod
+    def create(self) -> Container:
+        """Instantiate the image in a container."""
+
+    @abstractmethod
+    def wait(self, max_attempts: int = 15, sleep_rate: float = 1.0) -> bool:
+        """
+        Wait until the image is running.
+
+        :param max_attempts: max number of attempts.
+        :param sleep_rate: the amount of time to sleep between different requests.
+        :return: True if the wait was successful, False otherwise.
+        """
+        return True
+
+
+class GanacheDockerImage(DockerImage):
```

Comment:
Ganache should not be tied to aea base. This also is better placed in `aea_etherem_ledger.test_tools`.